// @edgepi-cloud/edgepi-rpc/lib/services/DoutService/DoutPins.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DoutPins = void 0;
const protobuf = __importStar(require("protobufjs"));
const root = protobuf.loadSync(`${__dirname}../../../../protos/dout.proto`);
const protoEnum = root.lookupEnum('DoutPins').values;
const DoutPins = Object.freeze({
    DOUT1: protoEnum.DOUT1,
    DOUT2: protoEnum.DOUT2,
    DOUT3: protoEnum.DOUT3,
    DOUT4: protoEnum.DOUT4,
    DOUT5: protoEnum.DOUT5,
    DOUT6: protoEnum.DOUT6,
    DOUT7: protoEnum.DOUT7,
    DOUT8: protoEnum.DOUT8
});
exports.DoutPins = DoutPins;
