// @edgepi-cloud/edgepi-rpc/lib/services/LedService/LedPins.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LEDPins = void 0;
const protobuf = __importStar(require("protobufjs"));
const root = protobuf.loadSync(`${__dirname}../../../../protos/led.proto`);
const protoEnum = root.lookupEnum('LEDPins').values;
const LEDPins = Object.freeze({
    LED1: protoEnum.LED1,
    LED2: protoEnum.LED2,
    LED3: protoEnum.LED3,
    LED4: protoEnum.LED4,
    LED5: protoEnum.LED5,
    LED6: protoEnum.LED6,
    LED7: protoEnum.LED7,
    LED8: protoEnum.LED8
});
exports.LEDPins = LEDPins;
